File: AMF5/src/test/java/javaPlatform/EventListenerTest.java
package javaPlatform;

import amf.apicontract.client.platform.AMFBaseUnitClient;
import amf.apicontract.client.platform.WebAPIConfiguration;
import amf.core.client.platform.AMFResult;
import amf.core.client.platform.config.AMFEvent;
import amf.core.client.platform.config.AMFEventListener;
import amf.core.client.platform.config.AMFEventNames;
import amf.core.client.platform.config.StartingParsingEvent;
import amf.core.client.platform.model.document.BaseUnit;
import amf.core.client.platform.model.document.Document;
import amf.core.client.platform.model.domain.DomainElement;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class EventListenerTest {

    @Test
    public void parseAMFGraph() throws ExecutionException, InterruptedException {
        final SpyEventListener spy = new SpyEventListener();
        final AMFBaseUnitClient client = WebAPIConfiguration.WebAPI()
                .withEventListener(spy)
                .baseUnitClient();

        final AMFResult parseResult = client.parse("file://src/test/resources/examples/banking-api.raml").get();

        final BaseUnit model = parseResult.baseUnit();
        assertNotNull(model);
        assertTrue(parseResult.conforms());
        assert(spy.listenedTo.size() != 0);
    }

    private class SpyEventListener implements AMFEventListener {

        public List<StartingParsingEvent> listenedTo = new ArrayList<>();

        @Override
        public void notifyEvent(AMFEvent event) {
            if (event.name().equals(AMFEventNames.StartedParse())) {
                listenedTo.add((StartingParsingEvent) event);
            }
        }
    }
}

File: AMF5/src/test/scala/scalaPlatform/EventListenerTest.scala
package scalaPlatform

import amf.apicontract.client.scala.RAMLConfiguration
import amf.core.client.platform.config.AMFEventNames
import amf.core.client.scala.config.{AMFEvent, AMFEventListener, StartingParsingEvent}
import org.junit.Assert.{assertNotNull, assertTrue}
import org.junit.runner.RunWith
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.should
import org.scalatestplus.junit.JUnitRunner

import scala.collection.mutable

@RunWith(classOf[JUnitRunner])
class EventListenerTest extends AsyncFlatSpec with should.Matchers {

  "AMF" should "notify about certain events" in {
    val spy = new SpyEventListener()
    val client = RAMLConfiguration.RAML10()
      .withEventListener(spy)
      .baseUnitClient()

    client.parse("file://src/test/resources/examples/banking-api.raml").map { parseResult =>
      val model = parseResult.baseUnit
      assertNotNull(model)
      assertTrue(parseResult.conforms)
      assert(spy.listenedTo.nonEmpty)
    }
  }

  class SpyEventListener extends AMFEventListener {
    val listenedTo: mutable.ListBuffer[StartingParsingEvent] = new mutable.ListBuffer[StartingParsingEvent]

    override def notifyEvent(event: AMFEvent): Unit = {
      if (event.name == AMFEventNames.StartedParse) listenedTo.append(event.asInstanceOf[StartingParsingEvent])
    }
  }
}
